cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefreshService.java

```diff
@@ -0,0 +1,314 @@
+  private ScheduledExecutorService executor;
+  private final int batchSize;
+
+
+  /**
+   * Constructs a {@link NamespaceSourceControlMetadataRefreshService} instance.
+   */
+  @Inject
+  public NamespaceSourceControlMetadataRefreshService(CConfiguration cConf,
+      TransactionRunner transactionRunner,
+      SourceControlOperationRunner sourceControlOperationRunner,
+      NamespaceId namespaceId, NamespaceAdmin namespaceAdmin) {
+    this.transactionRunner = transactionRunner;
+    this.sourceControlOperationRunner = sourceControlOperationRunner;
+    this.bufferTime = cConf.getLong(SourceControlManagement.METADATA_REFRESH_BUFFER_SECONDS);
+    this.namespaceId = namespaceId;
+    this.namespaceAdmin = namespaceAdmin;
+    this.lastRefreshTime = new AtomicLong(0);
+    this.running = new AtomicBoolean(false);
+    this.batchSize = cConf.getInt(AppFabric.STREAMING_BATCH_SIZE);
+
+    // Adding a jitter (a random value between 0 and 10) to the runInterval because if services for all
+    // the namespaces ran at the same time, it may create availability issues in task workers.
+    long interval =
+        cConf.getLong(SourceControlManagement.METADATA_REFRESH_INTERVAL_SECONDS);
+    if (interval > 0) {
+      Random random = new Random();
+      this.runInterval = interval + random.nextInt(
+          MAX_JITTER_SECONDS + 1);
+    } else {
+      this.runInterval = interval;
+    }
+  }
+
+  private RepositorySourceControlMetadataStore getRepoSourceControlMetadataStore(
+      StructuredTableContext context) {
+    return RepositorySourceControlMetadataStore.create(context);
+  }
+
+  private NamespaceSourceControlMetadataStore getNamespaceSourceControlMetadataStore(
+      StructuredTableContext context) {
+    return NamespaceSourceControlMetadataStore.create(context);
+  }
+
+  @Override
+  protected final ScheduledExecutorService executor() {
+    executor = Executors.newSingleThreadScheduledExecutor(
+        Threads.createDaemonThreadFactory("source-control-metadata-refresh-service-for-namespace-"
+            + namespaceId.getNamespace()));
+    return executor;
+  }
+
+  private RepositoryTable getRepositoryTable(StructuredTableContext context)
+      throws TableNotFoundException {
+    return new RepositoryTable(context);
+  }
+
+
+  // TODO(CDAP-21020): Refactor this function
+  private RepositoryMeta getRepositoryMeta(NamespaceId namespace)
+      throws RepositoryNotFoundException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      RepositoryTable table = getRepositoryTable(context);
+      RepositoryMeta repoMeta = table.get(namespace);
+      if (repoMeta == null) {
+        throw new RepositoryNotFoundException(namespace);
+      }
+
+      return repoMeta;
+    }, RepositoryNotFoundException.class);
+  }
+
+  // TODO(CDAP-21017): Optimize periodic refresh of source control metadata
+  @Override
+  public void runOneIteration() {
+    if (!running.compareAndSet(false, true)) {
```

Comment:
You shouldn't need an atomic boolean for this. The `AbstractScheduledService` should have the correct guarantee of how the scheduled run is.